### ISSUE TITLE
Analyzer crash when it sees or patterns in switch expression arms.

### ISSUE THREAD

**Issue by AraHaan** [SonarSource/sonar-dotnet#4209]
### Description

I have not reported this since about 4~5 months when the issue first started with my projects.

Basically every time I had a switch expression that utilized ``or`` inside of any of the arms it would throw a fit.

Stack Trace:
```
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.SymbolicExecution.SymbolicExecutionRunner' threw an exception of type 'SonarAnalyzer.SymbolicExecution.SymbolicExecutionException' with message 'Error processing method: Reset ## Method file: /Users/arahaan/Desktop/Sdk/zlib.managed/zlib.managed/InfBlocks.cs ## Method line: 79,8 ## Inner exception: System.NotSupportedException: OrPattern ##    at SonarAnalyzer.SymbolicExecution.CSharpExplodedGraph.VisitInstruction(ExplodedGraphNode node) ##    at SonarAnalyzer.SymbolicExecution.AbstractExplodedGraph.Walk() ##    at SonarAnalyzer.Rules.SymbolicExecution.SymbolicExecutionRunner.Analyze(CSharpExplodedGraph explodedGraph, SyntaxNodeAnalysisContext context) ##    at SonarAnalyzer.Extensions.SonarAnalysisContextExtensions.Analyze(CSharpSyntaxNode declarationBody, ISymbol symbol, Action`2 runAnalysis, SyntaxNodeAnalysisContext context) ## '. [/Users/arahaan/Desktop/Sdk/zlib.managed/zlib.managed.csproj]
```

### Repro steps

Please provide a code snippet that compiles which reproduces the problem. Alternatively, a link to a SonarCloud project where the issue appears, or a git repository, or attach a zip with a reproducer project.

Minimal preproducable code:
```cs
internal enum Example
{
    Item1,
    Item2,
    Item3,
    Item4,
    Item5,
}

internal class UseExample
{
    internal int ProcessExample(Example example)
    {
        return example switch
        {
            Example.Item1 or Example.Item2 => 0,
            Example.Item3 or Example.Item4 => 1,
            _ => 2,
        };
    }
}
```

### Known workarounds

Currently the only way is to either disable the analyzer (which is not intended because I want to be able to be notified of problematic code so it can be fixed asap) or to just ignore them.

### Related information

* C#/VB.NET Plugins version - I only use the nuget package analyzer so the plugins are not used (I think) when I do ``dotnet build``.
* Visual Studio version - 16.10p1 on windows, JetBrains Rider (latest version) used on my Mac.
* MSBuild / dotnet version MSBuild on both windows and my Mac is: ``16.10.0-preview-21126-01+6819f7ab0``, .NET on both is: ``6.0.100-preview.2.21155.3``
* SonarScanner for .NET version (if used) On my project I use ``*-*`` to automatically resolve to whatever the latest build of the analyzer is (because I got tired of seeing dependabot screwing up and downgrading the package on me plus it gives a free bonus of automatically giving me preview builds as well too if I add feeds that contains preview builds newer than the nuget.org releases). Currently that value is automatically resolved to ``8.20.0.28934``.
* Operating System - happens on my Windows 10 laptop with the latest dev build of windows 10, and my macbook with MacOS 11.3 Big Sur

(Note: Ironically it also seems that some ``if-else`` blocks in my code could be converted to normal switch and seems to crash for them as well and calling them ``RelationalPattern``s that can be split into a different issue however

```
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.SymbolicExecution.SymbolicExecutionRunner' threw an exception of type 'SonarAnalyzer.SymbolicExecution.SymbolicExecutionException' with message 'Error processing method: Bi_flush ## Method file: /Users/arahaan/Desktop/Sdk/zlib.managed/zlib.managed/Deflate.cs ## Method line: 730,8 ## Inner exception: System.NotSupportedException: RelationalPattern ##    at SonarAnalyzer.SymbolicExecution.CSharpExplodedGraph.VisitInstruction(ExplodedGraphNode node) ##    at SonarAnalyzer.SymbolicExecution.AbstractExplodedGraph.Walk() ##    at SonarAnalyzer.Rules.SymbolicExecution.SymbolicExecutionRunner.Analyze(CSharpExplodedGraph explodedGraph, SyntaxNodeAnalysisContext context) ##    at SonarAnalyzer.Extensions.SonarAnalysisContextExtensions.Analyze(CSharpSyntaxNode declarationBody, ISymbol symbol, Action`2 runAnalysis, SyntaxNodeAnalysisContext context) ## '. [/Users/arahaan/Desktop/Sdk/zlib.managed/zlib.managed.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.SymbolicExecution.SymbolicExecutionRunner' threw an exception of type 'SonarAnalyzer.SymbolicExecution.SymbolicExecutionException' with message 'Error processing method: Bi_windup ## Method file: /Users/arahaan/Desktop/Sdk/zlib.managed/zlib.managed/Deflate.cs ## Method line: 748,8 ## Inner exception: System.NotSupportedException: RelationalPattern ##    at SonarAnalyzer.SymbolicExecution.CSharpExplodedGraph.VisitInstruction(ExplodedGraphNode node) ##    at SonarAnalyzer.SymbolicExecution.AbstractExplodedGraph.Walk() ##    at SonarAnalyzer.Rules.SymbolicExecution.SymbolicExecutionRunner.Analyze(CSharpExplodedGraph explodedGraph, SyntaxNodeAnalysisContext context) ##    at SonarAnalyzer.Extensions.SonarAnalysisContextExtensions.Analyze(CSharpSyntaxNode declarationBody, ISymbol symbol, Action`2 runAnalysis, SyntaxNodeAnalysisContext context) ## '. [/Users/arahaan/Desktop/Sdk/zlib.managed/zlib.managed.csproj]
```
)

**Comment by pavel-mikula-sonarsource** [comment]
Hi @AraHaan,

Thank you for reporting this case. Luckily, it was fixed 12 days ago in #4104 on current master. The fix will be part of next release 8.21 / SQ 8.9.

**Comment by pavel-mikula-sonarsource** [comment]
Just to clarify the versioning in your case:
`C#/VB.NET Plugins version` - you're right, that the `plugin` itself isn't used. But the plugin version is always the same as the nuget version. They're released together. The automatically resolved version `8.20.0.28934` is the correct answer.
`SonarScanner for .NET version (if used)` - Scanner is not used in your case as you reference the nugets directly.

**Comment by AraHaan** [comment]
Thanks, so that covers the ``OrPattern``'s too when ``or`` is used in switch expression arms or in normal switch cases? 👌

**Comment by pavel-mikula-sonarsource** [comment]
The `OrPattern` will not throw, because it's not supported by our current CFG. Methods with `OrPattern` will not reach the Symbolic Execution engine where the `RelationalPattern` exception occurred.

**Comment by AraHaan** [comment]
Can it be possible for you guys to setup a feed for us to be able to add to our ``Nuget.config`` if we want to be able to automatically use the packages built from the tip of the default branch that has not been released yet to nuget.org?

So that way the issues that do get fixed before then they could try them out without needing to wait possibly a month or so to be able to use the fixes.

**Comment by pavel-mikula-sonarsource** [comment]
Unfortunately not for several reasons:
* Latest master is not guaranteed to be in a releasable state. There could be some WIP or a feature waiting for final validation.
* Without the last validation step, you can actually see more AD0001 warnings.
* Our usual release cycle is only 3 weeks long. And you've mentioned that this issue appeared ~4 months ago.

On the other hand, our repository is open source and you can build your own nuget for this purpose. With all the risks that comes with it.

**Comment by AraHaan** [comment]
@pavel-mikula-sonarsource what is the current status on that change? Is it now close to being released?

**Comment by pavel-mikula-sonarsource** [comment]
Hi @AraHaan, we'd like to release it today. We've already started, but there are some technical difficulties with the process in the toolchain we use.

**Comment by AraHaan** [comment]
What is the specific issue?
